### DISCUSSION TITLE
Database Performance

### DISCUSSION THREAD

**Discussion by blyles** [concourse/concourse#8454] [General]
![image](https://user-images.githubusercontent.com/5386554/175904049-fba04b36-a738-4598-a707-7ce5b771597c.png)

Since 7.7.x, we've had a lot of database pressures on 9 out of 17 installations (the bigger installations), we have 16vCPU and 64GB assigned to these.  You can see the sporadic cpu pressure every 60-90 seconds.

We've attempted the following;

- Followed all basic recommendations on tuning guides for GC.
- Tuned postgres shared_buffers, etc.  Memory management controls.

We're stepping it up to getting some further agents in there now for deeper introspection to the database but wanted to see if there's a general go to guide for this, as we'd like to avoid a 32 vCPU database to just cope with the spikes.

**Comment by blyles** [comment]
This ones stands out in our query plan execution - as a significant cost on rows.

Delete on check_build_events  (cost=11229.25..24225.66 rows=26856 width=6)
  CTE resource_builds
    ->  Append  (cost=4661.98..10993.22 rows=10490 width=8)
          ->  HashAggregate  (cost=4661.98..4714.93 rows=5295 width=8)
                Group Key: resource_config_scopes.last_check_build_id
                ->  Seq Scan on resource_config_scopes  (cost=0.00..4648.07 rows=5565 width=8)
                      Filter: (last_check_build_id IS NOT NULL)
          ->  HashAggregate  (cost=6068.99..6120.94 rows=5195 width=8)
                Group Key: resources.in_memory_build_id
                ->  Seq Scan on resources  (cost=0.00..6054.51 rows=5791 width=8)
                      Filter: (in_memory_build_id IS NOT NULL)
  ->  Seq Scan on check_build_events  (cost=236.03..13232.44 rows=26856 width=6)
        Filter: (NOT (hashed SubPlan 2))
        SubPlan 2
          ->  CTE Scan on resource_builds  (cost=0.00..209.80 rows=10490 width=8)

I dropped in the discord this query plan for resource checks and I've noticed this-  https://github.com/concourse/concourse/pull/8500

Looks like a huge fix for us and we're eager to get it in, is 7.9 coming soon? We're still seeing significant resource costs here.

**Comment by evanchaoli** [comment]
@blyles I can give you a workaround without waiting for 7.9, actually I am doing so on our clusters.

Just stop the check collector by:

```sql
update components set paused=true where name = 'collector_checks'
```

Then run a cronjob every minute, or a `time` triggered pipeline job to run the following two SQLs to cleanup check logs:

```sql
WITH resource_builds AS (
        SELECT distinct(last_check_build_id) as build_id
        FROM resource_config_scopes
        WHERE last_check_build_id IS NOT NULL
      ),
      deleted_builds AS (
        DELETE FROM builds USING (
          (SELECT id
          FROM builds b
          WHERE completed AND resource_id IS NOT NULL
          AND NOT EXISTS ( SELECT 1 FROM resource_builds WHERE build_id = b.id )
					LIMIT 500)
            UNION ALL
          SELECT id
          FROM builds b
          WHERE completed AND resource_type_id IS NOT NULL
          AND EXISTS (SELECT * FROM builds b2 WHERE b.resource_type_id = b2.resource_type_id AND b.id < b2.id)
    ) AS deletable_builds WHERE builds.id = deletable_builds.id
      RETURNING builds.id
      ), deleted_events AS (
        DELETE FROM check_build_events USING deleted_builds WHERE build_id = deleted_builds.id
      )
      SELECT COUNT(*) FROM deleted_builds
```

```sql
with expired_imb_ids as (
	select distinct(build_id) as build_id from check_build_events cbe 
		where not exists (select 1 from resources where in_memory_build_id = cbe.build_id) and 
		      not exists (select 1 from builds where id = cbe.build_id) 
)
delete from check_build_events cbe2 using expired_imb_ids where cbe2.build_id = expired_imb_ids.build_id;
```

In practice, we run the two SQLs in two different transactions. You may setup the cronjob first, then stop `collector_checks`.

**Reply by blyles** [reply]
This will work well actually - it seems this is our DB runaway usage based on our explain plans so gonna run this through one of our test systems.  Appreciate that, alot.
